### PR TITLE
feat: extract payments on club endpoint; feat: download attendance report; feat: update user's response to an event

### PR DESCRIPTION
**Pull request by ptmminh**
This feature is needed for our club. However, this is a new major addition to the module as a whole (new API endpoint club).

Related to #70 as far as I can tell.

Therefore, this is just a draft to see how @Olen and @elliot-100 feel about this new feature and start a conversation around it!

I can add some example script/usage if interested

### CODE REVIEW

**Comment by Olen** [comment]
In principle, I am all for extending the functionalty.
I have not tested the "club" functions, but if it is possible to unite the functions from the two endpoints without to many conflicting names I would prefer to keep it as one common class.
If not, using something like SpondClub or something is probably a good solution.

**Comment by elliot-100** [comment]
Very interesting! I keep meaning to look at Club functions for my own club.

For now - does this imply that other API endpoints (and the rest of the library) still work for you, but the club API endpoint provides additional functions?

I think that is what I was wondering in #70 ...

**Comment by ptmminh** [comment]
> For now - does this imply that other API endpoints (and the rest of the library) still work for you, but the club API endpoint provides additional functions?

Correct. The current library still works perfectly (i guess same functionality as provided in the front end spond.com/client). This new club endpoint is for the Spond Club functionality

**Comment by ptmminh** [comment]
> I have not tested the "club" functions, but if it is possible to unite the functions from the two endpoints without to many conflicting names I would prefer to keep it as one common class.
> If not, using something like SpondClub or something is probably a good solution.

Yeah of course i would prefer that too... but I think since the two front ends are different (spond.com/client vs. club.spond.com) i thought maybe having two classes might be a clearer separation of concern since the functionality covered in the two are also different but of course I'm happy and open to suggestions! I'll continue to have a think on it... The different name is also not a bad idea either if we are ok with two separate classes.

**Comment by elliot-100** [comment]
> I think since the two front ends are different (spond.com/client vs. club.spond.com) i thought maybe having two classes might be a clearer separation of concern

Agree, I think best to have them separate to start with, especially as it's not necessarily clear how they might functionally overlap or behave similarly-but-different (e.g. differing data structures). Seems safer to develop separately; common elements can always be refactored out to e.g. a base class.

**Comment by ptmminh** [comment]
I took a stab... Having both of them in the same class is difficult since we'll need two separate tokens to authenticate! Therefore, I opted for 2 separate classes and refactor out a base class. 

manual test script is going through! My downstream scripts seem to also work!

**Comment by ptmminh** [comment]
This PR ended up containing more features than originally intended. :)

**Comment by elliot-100** [comment]
Hi @ptmminh - many thanks for this. Could I ask you to run `isort .` after each commit and refresh with a force-push? This should reduce unnecessary diffs on merge later. It doesn't look like `black .` is needed. I will add an issue to add a pre-commit hook for convenience when developing.

**Comment by ptmminh** [comment]
@elliot-100 thanks for flagging the isort check. All done!

**Comment by ptmminh** [comment]
Just took a quick look at #114 too, do I also need to update README.md with additional features (`get_export`, `get_transactions` and `change_response`)?

**Comment by Olen** [comment]
Yes, please update the README.
Also, since the PR contains quite some refactoring of thr base class, I think we should look into merging a few of the other pending PRs first, so we only need to make any changes required for the refactoring once.

**Comment by ptmminh** [comment]
> Yes, please update the README.

Sure, done! :heavy_check_mark: 

> Also, since the PR contains quite some refactoring of thr base class, I think we should look into merging a few of the other pending PRs first, so we only need to make any changes required for the refactoring once.

Ok, feel free to merge other PRs first and then ping me and I'll rebase/refactor and we can merge it then.

**Comment by elliot-100** [comment]
> Ok, feel free to merge other PRs first and then ping me and I'll rebase/refactor and we can merge it then.

Other PRs merged now @ptmminh . I think I would have a couple of review comments but have run out of time today.

**Comment by ptmminh** [comment]
Done rebasing on top of current `main`

File: spond/spond.py [hunk 1]
```diff
@@ -1,74 +1,32 @@
...
 
 
-class AuthenticationError(Exception):
-    pass
-
-
-class Spond:
+class Spond(SpondBase):
 
-    API_BASE_URL = "https://api.spond.com/core/v1/"
```
Comment (elliot-100) [inline spond/spond.py:18]: Dropped my API_BASE_URL class constant refactor here

Comment (elliot-100) [reply spond/spond.py:18]: I already did that here: https://github.com/Olen/Spond/commit/9391ef278578e7b6b50ab0e7463a401162c751e8

You are including it in both `__super__()` calls so it should (remain) a base class constant.

(Apologies for any terseness, don't have much time today)

File: spond/club.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class SpondClub(SpondBase):
+    def __init__(self, username, password):
+        super().__init__(username, password, "https://api.spond.com/club/v1/")
+        self.transactions = None
+
+    @SpondBase.require_authentication
+    async def get_transactions(
+        self, club_id: str, skip: int = None, max_items: int = 100
```
Comment (elliot-100) [inline spond/club.py:13]: `skip` is None by default, so type hint is wrong - should be `Optional [int]` or `int | None` - former matches rest of codebase style

File: spond/spond.py (hunk 1 above)
Comment (ptmminh) [reply spond/spond.py:18]: Ok got it, I think i know what you mean, i won't get to it until today's evening just FYI

File: README.md [hunk 3]
```diff
@@ -57,6 +57,9 @@ Get all your messages.
 Send a message with content `text`.
 Either specify an existing `chat_id`, or both `user` and `group_uid` for a new chat.
 
+### get_export()
+Get excel report for an event attendance, available via the web client.
```
Comment (elliot-100) [inline README.md:61]: I think this should be something like "Get Excel attendance report for a single event..."

File: manual_test_functions.py [hunk 4]
```diff
@@ -40,6 +41,17 @@ async def main() -> None:
     for i, message in enumerate(messages):
         print(f"[{i}] {_message_summary(message)}")
 
+    # ATTENDANCE EXPORT
+
+    print("\nGetting a random attendance report...")
```
Comment (elliot-100) [inline manual_test_functions.py:46]: Nitpicking, but it's for the first event returned.

File: spond/spond.py [hunk 5]
```diff
@@ -378,3 +378,22 @@ async def update_event(self, uid, updates: dict):
         ) as r:
             self.events_update = await r.json()
             return self.events
+
+    @SpondBase.require_authentication
+    async def get_export(self, uid: str) -> bytes:
```
Comment (elliot-100) [inline spond/spond.py:383]: Think this is better named something more specific like `get_event_attendance_xlsx()` @Olen?

**Comment by elliot-100** [comment]
I don't have a Spond Club yet, but `manual_test_functions` works fine once I updated `config.py` with a dummy `club_id` - the attendance xlsx feature looks useful, didn't actually know that existed.

File: spond/spond.py (hunk 1 above)
Comment (ptmminh) [reply spond/spond.py:18]: Ok so spond API url is https://api.spond.com/core/v1/ and the club API is https://api.spond.com/club/v1/ 

They're different so i don't think it should be a base class constant?

Comment (ptmminh) [reply spond/spond.py:18]: or did i miss your point completely?

File: spond/club.py (hunk 2 above)
Comment (ptmminh) [reply spond/club.py]: done!

File: spond/club.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            and specify `max_items` instead. This param is only here for the
+            recursion implementation
+
+        Returns
+        -------
+        list of dict
+            A list of transactions, each represented as a dictionary.
+        """
+        if self.transactions is None:
+            self.transactions = list()
```
[the remark this answers is not in the document]
Comment (ptmminh) [reply spond/club.py]: done

File: spond/club.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        url = f"{self.api_url}transactions"
+        params = None if skip is None else {"skip": skip}
+        headers = {**self.auth_headers, "X-Spond-Clubid": club_id}
+
+        async with aiohttp.ClientSession() as session:
```
[the remark this answers is not in the document]
Comment (ptmminh) [reply spond/club.py]: actually, i refactored it to follow the rest of the codebase convention, two nested `with` for context managers is not necessary

File: spond/club.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            async with session.get(url, headers=headers, params=params) as response:
+                if response.status == 200:
```
[the remark this answers is not in the document]
Comment (ptmminh) [reply spond/club.py]: removed but good to know!

File: spond/base.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import aiohttp
+
+
+class AuthenticationError(Exception):
+    pass
+
+
+class SpondBase:
```
[the remark this answers is not in the document]
Comment (ptmminh) [reply spond/base.py]: done, good suggestion!

File: README.md (hunk 3 above)
Comment (ptmminh) [reply README.md]: Yeah i like that, more explicit, done!

File: manual_test_functions.py (hunk 4 above)
Comment (ptmminh) [reply manual_test_functions.py]: True, i adjusted the comment

File: spond/spond.py (hunk 5 above)
Comment (ptmminh) [reply spond/spond.py]: yep more explicit i changed it

File: spond/club.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ptmminh) [reply spond/club.py]: actually, please see [bbb6d07](https://github.com/Olen/Spond/pull/107/commits/bbb6d073ffd481fdbf972bf8902c1b0ab3003f4c)

**Comment by ptmminh** [comment]
Hmm still some `isort` linting error, i'll get to it later since i ran out of time

File: spond/spond.py (hunk 1 above)
Comment (elliot-100) [reply spond/spond.py:18]: Oops, sorry, misread!

File: spond/club.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (elliot-100) [reply spond/club.py]: Fair enough, and I can't test anyway (other than, as expected, no transactions are returned if I'm using a non-existent (dummy) Club ID)

**Review by elliot-100 — APPROVED** [review]
Many thanks for edits, all resolved and OK to merge as far as I am concerned.

**Comment by ptmminh** [comment]
yep, all set, please rebase or merge onto main! :rocket: 

Thanks for the review @elliot-100 ! :100:

**Comment by elliot-100** [comment]
@Olen any comments/changes? I can merge if you'd like.

**Comment by Olen** [comment]
Looks good to me

**Pull request merged**